### PR TITLE
Add support for fiat invoices

### PR DESCRIPTION
**Pull request by niteshbalusu11**
Added support for fiat invoices

### CODE REVIEW

File: commands/decode_tokens.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+const hasFiat = n => /(usd)/gim.test(n);
+const {isInteger} = Number;
+const isNumber = n => !isNaN(n);
+const symbols = ['USD'];
+const tokensAsBigTokens = n => Math.round(n * 1e8);
+
+/** Parse tokens
+
+  {
+    request: <NodeJs Request Function>
```
Comment (alexbosworth) [inline commands/decode_tokens.js]:
```suggestion
    request: <Request Function>
```

File: commands/decode_tokens.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+          return cbk(null, {rate: fiatRate(fiat.rate), tokens: tokensAsBigTokens(amount)});
+        }
+
+        // Exit early when the amount cannot be parsed as tokens
+        if (!isNumber(tokens)) {
+          return cbk(null, {error: true});
+        }
+
+        // Exit early when the amount is fractional
+        if (!isInteger(Number(tokens))) {
```
Comment (alexbosworth) [inline commands/decode_tokens.js]: Shouldn't fractional amounts be handled?

File: commands/handle_invoice_command.js [hunk 3]
```diff
@@ -23,11 +25,12 @@ const splitArguments = n => n.split(' ');
       from: <Node Name String>
       lnd: <Authenticated LND gRPC API Object>
     }]
+    request: <NodeJs Request Function>
```
Comment (alexbosworth) [inline commands/handle_invoice_command.js]:
```suggestion
    request: <Request Function>
```

File: commands/decode_tokens.js (hunk 2 above)
Comment (alexbosworth) [reply commands/decode_tokens.js]: Oh nm I see why they aren't

File: commands/decode_tokens.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+      }],
+
+      // Parse tokens
+      parseTokens: ['getFiatPrice', ({getFiatPrice}, cbk) => {
+        // Exit early if no tokens are provided
+        if (!tokens) {
+          return cbk(null, {tokens: defaultTokens});
+        }
+
+        if (!!getFiatPrice) {
```
Comment (alexbosworth) [inline commands/decode_tokens.js]:
```suggestion
        // Exit early when using a fiat denominated invoice
        if (!!getFiatPrice) {
```

File: replies/update_invoice_from_reply.js [hunk 5]
```diff
@@ -25,11 +26,12 @@ const split = n => n.split('\n');
       lnd: <Authenticated LND API Object>
       public_key: <Node Identity Public Key Hex String>
     }]
+    request: <NodeJs Request Function>
```
Comment (alexbosworth) [inline replies/update_invoice_from_reply.js]:
```suggestion
    request: <Request Function>
```

File: replies/update_invoice_from_reply.js [hunk 6]
```diff
@@ -95,7 +101,7 @@ module.exports = ({api, ctx, id, nodes}, cbk) => {
       }],
 
       // Details for the updated new invoice
-      updated: ['details', 'type', ({details, type}, cbk) => {
+      updated: ['details', 'type', async ({details, type}) => {
         // Exit early when the type of update is not known
         if (!type) {
           return cbk();
```
Comment (alexbosworth) [inline replies/update_invoice_from_reply.js]: cbk is no longer defined?

File: replies/update_invoice_from_reply.js [hunk 7]
```diff
@@ -137,19 +136,18 @@ module.exports = ({api, ctx, id, nodes}, cbk) => {
...
           return await ctx.reply(failure.message, failure.actions);
         }
 
         // Exit early when the amount is fractional
-        if (!!updated.is_fractional_amount) {
+        if (updated.is_fraction_error) {
           const failure = failureMessage({is_fractional_amount: true});
 
-          return await ctx.reply(failure.message, failure.actions);
+          return await ctx.reply(failure.message, failure.actions) && null;
```
Comment (alexbosworth) [inline replies/update_invoice_from_reply.js]: && null is that required?

**Comment by niteshbalusu11** [comment]
For parsing amount using hot-formula-parser, should I pass the parse_amount function from bos repo or install the dependency on ln-telegram and reacreate the file?

**Comment by alexbosworth** [comment]
I think recreate the dependency, a todo for later is to abstract that to a third repo

**Comment by niteshbalusu11** [comment]
Added amount parsing using hot-formula-parser, it now supports usd, eur, btc, m, k etc

**Comment by alexbosworth** [comment]
Cool i'll take a look

**Comment by alexbosworth** [comment]
still working on the merge

**Comment by alexbosworth** [comment]
I'm in the testing phase now

**Comment by niteshbalusu11** [comment]
Did it turn out to be complex or something or you had to make a lot of changes?

**Comment by alexbosworth** [comment]
There's a lack of tests really so that's what makes it take a while, i was restructuring things but then i figured out the way i was restructuring wan't going to work

**Comment by niteshbalusu11** [comment]
Oh okay. Cool.

**Pull request merged**